Add EMNISTDataModule

sugatoray opened the pull request:
## What does this PR do?
Closes #672, #676 and #685.

<details>
<summary>
<strong>A summary of changes and modifications</strong> :star: :fire: [CLICK TO EXPAND]
</summary>
<p>

- File **`Added`**: 
  - [x] `pl_bolts/datasets/emnist_dataset.py` :green_circle: 
  - [x] Contents:
      - [x] EMNIST_METADATA
      - [x] EMNIST dataset
      - [x] BinaryEMNIST dataset **`Need New PR or add to #672`** :warning: 

- File **`Added`**: 
  - [x] `pl_bolts/datamodules/emnist_dataset.py` :green_circle: 
  - [x] Contents:
    - [x] EMNISTDataModule
    - [x] BinaryEMNISTDataModule **`Need New PR or add to #672`** :warning:  

- Files **`Modified`**

  - **Package**: `pl_bolts`
    - [x] `pl_bolts/datasets/__init__.py` :green_circle: 
    - [x] `pl_bolts/datamodules/__init__.py` :green_circle: 
  
  - **Tests**: 
    - For `datamodules`:
      - [x] `tests/datamodules/test_imports.py` :green_circle: 
      - [x] `tests/datamodules/test_datamodules.py` `WIP` :orange_circle: 


> Adding `BinaryEMNIST` and `BinaryEMNISTDataModule` was logical, looking at how `MNIST` and `BinaryMNIST` (dataset and datamodules) were implemented.

</p>
</details>

<details>
<summary>About the dataset</summary>

![image](https://user-images.githubusercontent.com/10201242/123559103-ada3f300-d75f-11eb-8926-84f130e81113.png)
source: https://arxiv.org/pdf/1702.05373.pdf [Table-I]

![image](https://user-images.githubusercontent.com/10201242/123557894-15a30b00-d759-11eb-93d6-146ff9217b52.png)
source: https://arxiv.org/pdf/1702.05373.pdf [Table-II]

</details>

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements) #672 
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/lightning-bolts/blob/master/.github/CONTRIBUTING.md), Pull Request section? `Y` :green_circle: 
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? `Y` :green_circle: 
- [x] Did you make sure to update the documentation with your changes? `Y` :green_circle: 
- [x] Did you write any new necessary tests? [not needed for typos/docs] `Y` :green_circle: 
- [x] Did you verify new and existing tests pass locally with your changes? `Y` :green_circle: 
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/lightning-bolts/blob/master/CHANGELOG.md)? `Y` :green_circle: 

<!-- For CHANGELOG separate each item in unreleased section by a blank line to reduce collisions -->

## PR review
 - [x] Is this pull request ready for review? (if not, please submit in draft mode) **`READY`** :green_circle: 

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

pep8speaks commented:
Hello @sugatoray! Thanks for opening this PR. 



* In the file [`pl_bolts/datamodules/binary_emnist_datamodule.py`](https://github.com/PyTorchLightning/lightning-bolts/blob/c8dec9a1d28d65c88bd36f79e9b6f070fe961a44/pl_bolts/datamodules/binary_emnist_datamodule.py):

> [Line 21:47](https://github.com/PyTorchLightning/lightning-bolts/blob/c8dec9a1d28d65c88bd36f79e9b6f070fe961a44/pl_bolts/datamodules/binary_emnist_datamodule.py#L21): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 24:37](https://github.com/PyTorchLightning/lightning-bolts/blob/c8dec9a1d28d65c88bd36f79e9b6f070fe961a44/pl_bolts/datamodules/binary_emnist_datamodule.py#L24): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 61:81](https://github.com/PyTorchLightning/lightning-bolts/blob/c8dec9a1d28d65c88bd36f79e9b6f070fe961a44/pl_bolts/datamodules/binary_emnist_datamodule.py#L61): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 62:70](https://github.com/PyTorchLightning/lightning-bolts/blob/c8dec9a1d28d65c88bd36f79e9b6f070fe961a44/pl_bolts/datamodules/binary_emnist_datamodule.py#L62): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 63:79](https://github.com/PyTorchLightning/lightning-bolts/blob/c8dec9a1d28d65c88bd36f79e9b6f070fe961a44/pl_bolts/datamodules/binary_emnist_datamodule.py#L63): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 70:83](https://github.com/PyTorchLightning/lightning-bolts/blob/c8dec9a1d28d65c88bd36f79e9b6f070fe961a44/pl_bolts/datamodules/binary_emnist_datamodule.py#L70): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 96:5](https://github.com/PyTorchLightning/lightning-bolts/blob/c8dec9a1d28d65c88bd36f79e9b6f070fe961a44/pl_bolts/datamodules/binary_emnist_datamodule.py#L96): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)
> [Line 102:18](https://github.com/PyTorchLightning/lightning-bolts/blob/c8dec9a1d28d65c88bd36f79e9b6f070fe961a44/pl_bolts/datamodules/binary_emnist_datamodule.py#L102): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment
> [Line 105:5](https://github.com/PyTorchLightning/lightning-bolts/blob/c8dec9a1d28d65c88bd36f79e9b6f070fe961a44/pl_bolts/datamodules/binary_emnist_datamodule.py#L105): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)
> [Line 109:41](https://github.com/PyTorchLightning/lightning-bolts/blob/c8dec9a1d28d65c88bd36f79e9b6f070fe961a44/pl_bolts/datamodules/binary_emnist_datamodule.py#L109): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 110:57](https://github.com/PyTorchLightning/lightning-bolts/blob/c8dec9a1d28d65c88bd36f79e9b6f070fe961a44/pl_bolts/datamodules/binary_emnist_datamodule.py#L110): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 112:57](https://github.com/PyTorchLightning/lightning-bolts/blob/c8dec9a1d28d65c88bd36f79e9b6f070fe961a44/pl_bolts/datamodules/binary_emnist_datamodule.py#L112): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 114:1](https://github.com/PyTorchLightning/lightning-bolts/blob/c8dec9a1d28d65c88bd36f79e9b6f070fe961a44/pl_bolts/datamodules/binary_emnist_datamodule.py#L114): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 123:1](https://github.com/PyTorchLightning/lightning-bolts/blob/c8dec9a1d28d65c88bd36f79e9b6f070fe961a44/pl_bolts/datamodules/binary_emnist_datamodule.py#L123): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 125:5](https://github.com/PyTorchLightning/lightning-bolts/blob/c8dec9a1d28d65c88bd36f79e9b6f070fe961a44/pl_bolts/datamodules/binary_emnist_datamodule.py#L125): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)
> [Line 128:42](https://github.com/PyTorchLightning/lightning-bolts/blob/c8dec9a1d28d65c88bd36f79e9b6f070fe961a44/pl_bolts/datamodules/binary_emnist_datamodule.py#L128): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace

* In the file [`pl_bolts/datamodules/emnist_datamodule.py`](https://github.com/PyTorchLightning/lightning-bolts/blob/c8dec9a1d28d65c88bd36f79e9b6f070fe961a44/pl_bolts/datamodules/emnist_datamodule.py):

> [Line 22:47](https://github.com/PyTorchLightning/lightning-bolts/blob/c8dec9a1d28d65c88bd36f79e9b6f070fe961a44/pl_bolts/datamodules/emnist_datamodule.py#L22): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 25:37](https://github.com/PyTorchLightning/lightning-bolts/blob/c8dec9a1d28d65c88bd36f79e9b6f070fe961a44/pl_bolts/datamodules/emnist_datamodule.py#L25): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 40:23](https://github.com/PyTorchLightning/lightning-bolts/blob/c8dec9a1d28d65c88bd36f79e9b6f070fe961a44/pl_bolts/datamodules/emnist_datamodule.py#L40): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment
> [Line 41:24](https://github.com/PyTorchLightning/lightning-bolts/blob/c8dec9a1d28d65c88bd36f79e9b6f070fe961a44/pl_bolts/datamodules/emnist_datamodule.py#L41): [E114](https://duckduckgo.com/?q=pep8%20E114) indentation is not a multiple of four (comment)
> [Line 41:24](https://github.com/PyTorchLightning/lightning-bolts/blob/c8dec9a1d28d65c88bd36f79e9b6f070fe961a44/pl_bolts/datamodules/emnist_datamodule.py#L41): [E116](https://duckduckgo.com/?q=pep8%20E116) unexpected indentation (comment)
> [Line 41:70](https://github.com/PyTorchLightning/lightning-bolts/blob/c8dec9a1d28d65c88bd36f79e9b6f070fe961a44/pl_bolts/datamodules/emnist_datamodule.py#L41): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 62:81](https://github.com/PyTorchLightning/lightning-bolts/blob/c8dec9a1d28d65c88bd36f79e9b6f070fe961a44/pl_bolts/datamodules/emnist_datamodule.py#L62): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 63:70](https://github.com/PyTorchLightning/lightning-bolts/blob/c8dec9a1d28d65c88bd36f79e9b6f070fe961a44/pl_bolts/datamodules/emnist_datamodule.py#L63): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 64:79](https://github.com/PyTorchLightning/lightning-bolts/blob/c8dec9a1d28d65c88bd36f79e9b6f070fe961a44/pl_bolts/datamodules/emnist_datamodule.py#L64): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 71:83](https://github.com/PyTorchLightning/lightning-bolts/blob/c8dec9a1d28d65c88bd36f79e9b6f070fe961a44/pl_bolts/datamodules/emnist_datamodule.py#L71): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 96:5](https://github.com/PyTorchLightning/lightning-bolts/blob/c8dec9a1d28d65c88bd36f79e9b6f070fe961a44/pl_bolts/datamodules/emnist_datamodule.py#L96): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)
> [Line 102:18](https://github.com/PyTorchLightning/lightning-bolts/blob/c8dec9a1d28d65c88bd36f79e9b6f070fe961a44/pl_bolts/datamodules/emnist_datamodule.py#L102): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment
> [Line 105:5](https://github.com/PyTorchLightning/lightning-bolts/blob/c8dec9a1d28d65c88bd36f79e9b6f070fe961a44/pl_bolts/datamodules/emnist_datamodule.py#L105): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)
> [Line 109:41](https://github.com/PyTorchLightning/lightning-bolts/blob/c8dec9a1d28d65c88bd36f79e9b6f070fe961a44/pl_bolts/datamodules/emnist_datamodule.py#L109): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 110:57](https://github.com/PyTorchLightning/lightning-bolts/blob/c8dec9a1d28d65c88bd36f79e9b6f070fe961a44/pl_bolts/datamodules/emnist_datamodule.py#L110): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 112:57](https://github.com/PyTorchLightning/lightning-bolts/blob/c8dec9a1d28d65c88bd36f79e9b6f070fe961a44/pl_bolts/datamodules/emnist_datamodule.py#L112): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 114:1](https://github.com/PyTorchLightning/lightning-bolts/blob/c8dec9a1d28d65c88bd36f79e9b6f070fe961a44/pl_bolts/datamodules/emnist_datamodule.py#L114): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 123:1](https://github.com/PyTorchLightning/lightning-bolts/blob/c8dec9a1d28d65c88bd36f79e9b6f070fe961a44/pl_bolts/datamodules/emnist_datamodule.py#L123): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 125:5](https://github.com/PyTorchLightning/lightning-bolts/blob/c8dec9a1d28d65c88bd36f79e9b6f070fe961a44/pl_bolts/datamodules/emnist_datamodule.py#L125): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)
> [Line 128:42](https://github.com/PyTorchLightning/lightning-bolts/blob/c8dec9a1d28d65c88bd36f79e9b6f070fe961a44/pl_bolts/datamodules/emnist_datamodule.py#L128): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace

* In the file [`pl_bolts/datasets/__init__.py`](https://github.com/PyTorchLightning/lightning-bolts/blob/c8dec9a1d28d65c88bd36f79e9b6f070fe961a44/pl_bolts/datasets/__init__.py):

> [Line 37:14](https://github.com/PyTorchLightning/lightning-bolts/blob/c8dec9a1d28d65c88bd36f79e9b6f070fe961a44/pl_bolts/datasets/__init__.py#L37): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace



Do see the [Hitchhiker's guide to code style](https://goo.gl/hqbW4r)

sugatoray commented:
| Split Name     | Classes | No. Training | No. Testing | Validation | Total   |
|----------|---------|--------------|-------------|------------|---------|
| By_Class | 62      | 697,932      | 116,323     | No   :warning:      | 814,255 |
| By_Merge | 47      | 697,932      | 116,323     | No   :warning:      | 814,255 |
| Balanced | 47      | 112,800      | 18,800      | Yes        | 131,600 |
| Digits   | 10      | 240,000      | 40,000      | Yes        | 280,000 |
| Letters  | 37      | 88,800       | 14,800      | Yes        | 103,600 |
| MNIST    | 10      | 60,000       | 10,000      | Yes        | 70,000  |

## Concern(s) about current implementation: :question: 

- The validation data does not exist for `byclass` and `bymerge` splits, as indicated in the [paper, *table-II*](https://arxiv.org/pdf/1702.05373.pdf). So, should there be a check for these `splits`, where if the user specifies to use no validation data, then the datamodule must not produce any validation dataset and datamodule?

  - one way this could be done is by introducing another keyword argument: `strict_val_split: bool = False` in the `__init__` method of `EMNISTDataModule` and then check if the metadata for that split has `validation = False`.

akihironitta commented on pl_bolts/datasets/emnist_dataset.py:
```diff
@@ -12,6 +12,75 @@
+
+"""
+
+EMNIST_METADATA = {
```
How will this `EMNIST_METADATA` be used? Is there any use case?

akihironitta replied:
> ## Concern(s) about current implementation: :question: 
> 
> - The validation data does not exist for `byclass` and `bymerge` splits, as indicated in the [paper, *table-II*](https://arxiv.org/pdf/1702.05373.pdf). So, should there be a check for these `splits`, where if the user specifies to use no validation data, then the code will not produce any validation dataset and datamodule?
> 
>   - one way this could be done is by introducing another keyword argument: `strict_val_split: bool = False` in the `__init__` method of `EMNISTDataModule` and then check if the metadata for that split has `validation = False`.

Sorry I missed your comment above.

akihironitta replied:
Just had a look at the paper. It says:
> Table II contains a summary of the EMNIST datasets and
indicates which classes contain a validation subset in the
training set. In these datasets, the last portion of the training
set, equal in size to the testing set, is set aside as a validation
set. Additionally, this subset is also balanced such that it
contains an equal number of samples for each task. If the
validation set is not to be used, then the training set can be
used as one contiguous set.

As far as I understand, it's just saying that "Validation" in the table means that datasets are balanced or not per class. I think we can just ignore the table and make training and validation sets from torchvision's train set, which is the same as other existing datasets like MNIST. no?

sugatoray replied:
@akihironitta Let me take a look at it again and get back to you on this one.

sugatoray replied:
I think, after taking another look at the table caption, where `validation = yes`, 

- `len(validation_dataset) == len(test_dataset) == N` 
- `train_dataset[-N:] == validation_dataset`
- balanced datasets: `val` and `test`

On the other hand, where `validation = no`, there is no specification on the size or "balancedness" of the validation dataset. I suggest that the user should be given an option to choose one from the following:

- Create train/test and validation datasets extactly as defined in the paper
- Create train/test and validation datasets to some specification: example -- `val_split = 0.2`

However, if necessary, this change could be introduced as a separate PR later (may be).

What do you think?

akihironitta replied:
That sounds reasonable! It'd be great if you could do it in this PR.

sugatoray replied:
@akihironitta This `val_split` logic has been implemented now for both `EMNISTDataModule` and `BinaryEMNISTDataModule`, along with the test.

akihironitta replied:
@sugatoray Could you remove unused keys, such as "num_total"?

akihironitta reviewed (commented):
@sugatoray This would be a great addition to Bolts! Could you have a look at my comments below?

akihironitta commented:
@sugatoray Would you mind updating the changelog as well?

sugatoray commented on pl_bolts/datasets/emnist_dataset.py:
```diff
@@ -0,0 +1,138 @@
+        )   # noqa: E501
+    ]
+
+EMNIST._metadata = EMNIST_METADATA
```
Thank you. Yes, this totally makes sense. Will update the code.

sugatoray commented on pl_bolts/datasets/emnist_dataset.py:
```diff
@@ -0,0 +1,138 @@
+}
+
+
+# TODO(sugatoray): This is needed to avoid 503 error when downloading MNIST dataset
+# from http://yann.lecun.com/exdb/mnist/ and can be removed after `torchvision==0.9.1`.
+# See https://github.com/pytorch/vision/issues/3549 for details.
+if _TORCHVISION_AVAILABLE and _TORCHVISION_LESS_THAN_0_9_1:
+    EMNIST.resources = [
+        (
+            "https://ossci-datasets.s3.amazonaws.com/mnist/train-images-idx3-ubyte.gz",
+            "f68b3c2dcbeaaa9fbdd348bbdeb94873"
+        ),  # noqa: E501
+        (
+            "https://ossci-datasets.s3.amazonaws.com/mnist/train-labels-idx1-ubyte.gz",
+            "d53e105ee54ea40749a09fcbcd1e9432"
+        ),  # noqa: E501
+        (
+            "https://ossci-datasets.s3.amazonaws.com/mnist/t10k-images-idx3-ubyte.gz",
+            "9fb629c4189551a2d022fa330f9573f3"
+        ),  # noqa: E501
+        (
+            "https://ossci-datasets.s3.amazonaws.com/mnist/t10k-labels-idx1-ubyte.gz",
+            "ec29112dd5afa0611ce80d1b7f02629c"
+        )   # noqa: E501
+    ]
```
@akihironitta Are you sure about it? If you import (from `torchvision`) and check the `EMNIST.resources` it gives you the same output as `MNIST.resources`. So, I am not sure if the problem is only limited to `MNIST`.

akihironitta replied:
@sugatoray Yes, I think I am.

> If you import (from `torchvision`) and check the `EMNIST.resources` it gives you the same output as `MNIST.resources`. So, I am not sure if the problem is only limited to `MNIST`.

`EMNIST.resources` is the same value because the class is a subclass of `MNIST`. If you look at the code (https://github.com/pytorch/vision/blob/ef711591a5db69d36f904ab5c39dec13627a58ad/torchvision/datasets/mnist.py#L256), `EMNIST` never uses `EMNIST.resources` for downloading the dataset.

sugatoray replied:
@akihironitta Yes, you are right. I checked the code as well. The `download()` method in `EMNIST` calls `download_and_extract_archive()` and DOES NOT use the `EMNIST.resources` attribute.

```python
download_and_extract_archive(self.url, download_root=self.raw_folder, md5=self.md5)
``` 

Where, 

```python
self.url = 'https://www.itl.nist.gov/iaui/vip/cs_links/EMNIST/gzip.zip'
self.md5 = "58c8d27c78d21e728a6bc7b3cc06412e"
```

sugatoray commented on tests/datamodules/test_datamodules.py:
```diff
@@ -85,3 +87,26 @@ def _create_dm(dm_cls, datadir, val_split=0.2):
     return dm
+
+
+# @pytest.mark.parametrize("datadir", ['./tempdata']) ## TODO: Remove this before final commit
```
@akihironitta Yes, this I had there to test the code outside the repo, whithout cloning the whole repo. I tried building the whole thing in Google colab, in a single notebook -- as an experiment. I might share my notes later on, so if others want to use colab for such dev-activities, what could be a clean simple (and nimble) process.

As for this commented line, I will remove it.

sugatoray commented:
> @sugatoray Would you mind updating the changelog as well?

@akihironitta Sure. However, am not sure which *changelog* to update? It will be helpful if you could point me in the right direction. Thank you.

akihironitta commented:
> > @sugatoray Would you mind updating the changelog as well?
> 
> @akihironitta Sure. However, am not sure which _changelog_ to update? It will be helpful if you could point me in the right direction. Thank you.

@sugatoray Thank you for having a look at my comments! The changelog is here :] https://github.com/PyTorchLightning/lightning-bolts/blob/master/CHANGELOG.md

sugatoray commented:
@akihironitta Here is a draft of the changelog. Does it look okay to you?

## [unReleased] - 2021-MM-DD

### Added
- Added `EMNISTDataModule`, `BinaryEMNISTDataModule`, `EMNIST` dataset, `BinaryEMNIST` dataset and `EMNIST_METADATA` ([#676](https://github.com/PyTorchLightning/lightning-bolts/pull/676)).

### Changed


### Deprecated


### Removed


### Fixed

- Fixed the MNIST download giving HTTP 404 with torchvision>=0.9.1 ([#674](https://github.com/PyTorchLightning/lightning-bolts/pull/674))


- Removed momentum updating from val step and add separate val queue ([#631](https://github.com/PyTorchLightning/lightning-bolts/pull/631))

akihironitta commented on pl_bolts/datamodules/emnist_datamodule.py:
```diff
@@ -0,0 +1,178 @@
+        self,
+        data_dir: Optional[str] = None,
+        split: str = 'mnist',
+        val_split: Union[int, float] = 0.2,
```
This is only a suggestion, but how about setting `val_split` to `None` by default, and in `__init__(...)`, assigning a `int` or `float` value defined in the paper to `val_split`?

```suggestion
        val_split: Optional[int, float] = None,
```

sugatoray replied:
@akihironitta `MNISTDataModule` and `BinaryMNISTDataModule` uses a `val_split = 0.2`. So, I suggest we keep the same as a default argument for `EMNISTDataModule` and `BinaryEMNISTDataModule`. This will allows for less stringent requirement for the user to visit documentation while using the API: *This will at least create train-val-test datasets, each of finite length.* 

> *Think of all the packages that follow sklearn's api format; a typical user, who has used sklearn's API, would know almost immediately the design of another similar API, even before using it*. 

However, I like your suggestion of using `None` as a value to flag the use of **default** train-val-test splitting of the EMNIST dataset. In fact, the current implementation, uses `val_split is None` as the conditional to fetch the validation dataset size (this requires the EMNIST_METADATA).

Additionally, I have added another layer of abstraction in `EMNISTDataModule` (I have not yet implemented the same for `BinaryEMNISTDataModule` though): 

```python
# _DEFAULT_NO_VALIDATION_VAL_SPLIT: This is the `val_split` to use when
# "validation = False" for a given split in the metadata and the user-input
# for `val_split` is `None`.
_DEFAULT_NO_VALIDATION_VAL_SPLIT: Union[int, float] = 0.
```

## About using `Optional` for `val_split`

If you want to use `Optional` for type-hinting `val_split`, I think you have a typo in what you suggested.  

```Diff
-        val_split: Optional[int, float] = None,
+        val_split: Optional[Union[int, float]] = 0.2,
```

The `Optional[...]` takes only a single type as argument, and so, if I understood you correctly, we could use either of the following (see [this stackoverflow discussion][#so-question] for more details):

[#so-question]: https://stackoverflow.com/questions/51710037/how-should-i-use-the-optional-type-hint

```python
# method-1: Optional + Union
val_split: Optional[Union[int, float]] = 0.2 # or, None

# method-2: Union + None
val_split: Union[int, float, None] = 0.2 # or, None
```

:star: :question: Which method you think is better: `method-1` or `method-2`?

akihironitta replied:
Sure. Let me check my understanding of the behaviour you're suggesting here. Are the following statements correct?:

- `val_split` is set to `0.2` by default to keep consistency with sklearn.
- Passing `None` to `val_split` uses the train-val split defined in the paper for the balanced splits.
- Passing `None` to `val_split` uses `0` (no validation set) for non-balanced splits.

---

As for the typing `Optional`, you're right. It's just a typo :sweat_smile:

akihironitta commented:
@sugatoray Could you allow us to edit your PR so that I can directly push some minor fixes/improvements to your branch? https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork

sugatoray commented on pl_bolts/datasets/__init__.py:
```diff
@@ -10,6 +10,7 @@
     RandomDictDataset,
     RandomDictStringDataset,
 )
+from pl_bolts.datasets.emnist_dataset import BinaryEMNIST, EMNIST
```
@akihironitta We need to import `EMNIST` from `pl_bolts.datasets.emnist_dataset` as we still need to add the `_metadata` attribute to it. The `EMNIST._metadata` is later on used in `EMNISTDataModule` and `BinaryEMNISTDataModule` to ensure the correct evaluation (determined dynamically based on the value of `split` keyword argument) for: 

- check if valid `split` is provided
- retirn correct `num_classes` for a given `split`
- correct `val_split` assignment for a given `split`

Besides, this keeps the implementation of `EMNIST : BinaryEMNIST` easily perceivable for someone who went through `MNIST : BinaryMNIST` and vice-versa. Also, if you ever need to add/update anything to `EMNIST` and thereby to `BinaryEMNIST`, the present implementation provides *a clear single point of entry* (in `pl_bolts.datasets.emnist_dataset.py`) for future development/maintenance purposes. 

Therefore, as I see it fit, `EMNIST` should not be imported from `torchvision.datasets.EMNIST` directly and I believe that the current implementation should suffice what is needed.

Does it make sense?

akihironitta replied:
> - check if valid `split` is provided
> - return correct `num_classes` for a given `split`

For these, please have a look at https://github.com/PyTorchLightning/lightning-bolts/pull/676#discussion_r665078561 and https://github.com/PyTorchLightning/lightning-bolts/pull/676#discussion_r665080491.

sugatoray commented:
> @sugatoray Could you allow us to edit your PR so that I can directly push some minor fixes/improvements to your branch? https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork

@akihironitta You have editor access the PR branch now.

![image](https://user-images.githubusercontent.com/10201242/124681326-e0e82f80-de8d-11eb-8b7c-cfc79e75fef0.png)

sugatoray commented:
Keeping this table handy here (just in case):

```python
"""
    Specs:
        - 6 splits: ``byclass``, ``bymerge``,
                    ``balanced``, ``letters``,
                    ``digits`` and ``mnist``.
        - Table:
          | Split Name   | Classes | No. Training | No. Testing | Validation | Total   |
          |--------------|---------|--------------|-------------|------------|---------|
          | ``byclass``  | 62      | 697,932      | 116,323     | No         | 814,255 |
          | ``bymerge``  | 47      | 697,932      | 116,323     | No         | 814,255 |
          | ``balanced`` | 47      | 112,800      | 18,800      | Yes        | 131,600 |
          | ``digits``   | 10      | 240,000      | 40,000      | Yes        | 280,000 |
          | ``letters``  | 37      | 88,800       | 14,800      | Yes        | 103,600 |
          | ``mnist``    | 10      | 60,000       | 10,000      | Yes        | 70,000  |
          source: https://arxiv.org/pdf/1702.05373.pdf [Table-II]
        - For each split:
          - Each image is (1 x 28 x 28)
"""
```

sugatoray commented:
@akihironitta I also added the documentation (for sphinx) under `docs/source/`. The datamodules with :star: were developed under this PR. But other datamodules (with only :green_circle: ) were just added to the documentation. 

:question: Perhaps I could open a separate issue to address this update (updating datamodules documentation generation with sphinx)?

- [x] File: `datamodules_others.rst`: 
  - [x] `BinaryEMNISTDataModule` :green_circle: :star:
  - [x] `BinaryMNISTDataModule` :green_circle:
  - [x] `CIFAR10DataModule` :green_circle:
  - [x] `CityscapesDataModule` :green_circle:
  - [x] `EMNISTDataModule` :green_circle: :star:
  - [x] `FashionMNISTDataModule` :green_circle:
  - [x] `ImagenetDataModule` :green_circle:
  - [x] `KittiDataModule` :green_circle:
  - [x] `MNISTDataModule` :green_circle:
  - [x] `SSLImagenetDataModule` :green_circle:
  - [x] `STL10DataModule` :green_circle:

- [x] File: `datasets_others.rst`

akihironitta commented:
@sugatoray Thanks for having a look at my branch (https://github.com/PyTorchLightning/lightning-bolts/tree/feature/672_EMNISTDataModule) and applying changes.

akihironitta commented on docs/source/datamodules_others.rst:
```diff
@@ -0,0 +1,69 @@
+.. role:: hidden
```
Why do we need this? We already have https://lightning-bolts.readthedocs.io/en/latest/datamodules_vision.html.

sugatoray replied:
@akihironitta Thank you, for pointing it out. When I looked at the documentation, I could not find the modules (only Vision and Sklearn modules were there) on the navigation pane. Is there a way to make the modules appear/collapse on the navigation pane?

I will add `EMNISTDataModule` and `BinaryEMNISTDataModule` to `datamodules_vision.rst` and remove `datamodules_others.rst`.

sugatoray replied:
:bulb: The latest update is posted above [here](https://github.com/PyTorchLightning/lightning-bolts/pull/676#issuecomment-875279745).

akihironitta commented on docs/source/datasets_others.rst:
```diff
@@ -0,0 +1,26 @@
+.. role:: hidden
```
Same here. Please have a look at my comment above.

sugatoray replied:
@akihironitta Thank you. I will remove this one as well.

akihironitta commented on pl_bolts/datamodules/binary_emnist_datamodule.py:
```diff
@@ -0,0 +1,180 @@
+            else:
+                self.val_split = self._DEFAULT_NO_VALIDATION_VAL_SPLIT
+
+    @property
+    def _emnist_split_exists(self) -> bool:
+        """Checks if the split exists in emnist or not."""
+        if not any(s == self.split for s in self.dataset_cls.splits):
+            allowed_values = ', '.join([f'"{v}"' for v in self.dataset_cls.splits])
+            raise ValueError(
+                f'Invalid value provided for split (="{self.split}"). ' + f'Allowed splits are: {allowed_values}'
+            )
+        return True
+
```
I don't think we need to check if a given `split` is valid as torchvision does it anyway at https://github.com/pytorch/vision/blob/015378434c1e4be778b2d65eb86f227db20bc8bf/torchvision/datasets/mnist.py#L291
```suggestion
```

sugatoray replied:
@akihironitta You are correct that torchvision does run a check. It came to my mind while writing this method, but I kept it there as an additional fail-safe. See, the dataset is not called before the `prepare()` and then the `setup()` methods. These methods are also not called at the time of instantiation of the datamodule class. So, technically, one could check `num_classes` property even before calling (fetching) the dataset. The `_emnist_split_exists()` ensures that even in such cases, the code does not do something odd.

sugatoray replied:
:bulb: Given that the dependency of `num_classes` has been lifted from EMNIST_METADATA, I think this method `_emnist_split_exists()` could be removed as well.

akihironitta replied:
@sugatoray To me, it' looks just redundant since it's something that's been done already in torchvision. It's up to you then, but other maintainers might suggest removing it later.

sugatoray replied:
@akihironitta Thank you. I am keeping this as a non-resolved conversation to get back to this later on. Thank you, for all your reviews. Let me attend to them first, and then we can decide what to do about this later.

akihironitta commented on pl_bolts/datamodules/binary_emnist_datamodule.py:
```diff
@@ -0,0 +1,180 @@
+        """
+        # The _metadata is only added to EMNIST dataset
+        # to get split-specific metadata.
+        return self.split_metadata.get('num_classes')
```
Shall we utilise torchvision's attribute here so that we can get rid of a metadata key `EMNIST_METADATA["splits"][...]["num_classes"]`?https://github.com/pytorch/vision/blob/015378434c1e4be778b2d65eb86f227db20bc8bf/torchvision/datasets/mnist.py#L281
```suggestion
        return len(self.dataset_cls.classes_split_dict[self.split])
```

sugatoray replied:
@akihironitta The EMNIST dataset class (from torchvision) does not provide an easy way to access all the content as in Table-II of the paper. I believe, it is very helpful in making it available as a `dict` to the user. This let's you quickly check different aspects of the dataset. More importantly, EMNIST_METADATA is used to run some tests and determine default-validation dataset size for a split.

About using the `classes_split_dict`, yes that can be done (as you showed) and perhaps making `num_classes` dependent only on the core `EMNIST` attributes is a better idea, as this is more organic.

```python
from torchvision.datasets import EMNIST
dict((k, len(v)) for k, v in EMNIST.classes_split_dict.items())
# {'balanced': 47,
#  'byclass': 62,
#  'bymerge': 47,
#  'digits': 10,
#  'letters': 27,
#  'mnist': 10}
```

:warning: However, I would still advise to retain the EMNIST_METADATA as it is, since, the length of the default validation data is determined based on the test data and the split. The as-is version of EMNIST does not provide this information (AFAIK). I added the EMNIST_METADATA as an easy means to access Table-II as well.

akihironitta reviewed (commented):
@sugatoray Thank you again for your contribution! I'm sorry but I haven't look at all of your code yet. I'll have another look later as soon as possible.

akihironitta commented on pl_bolts/datamodules/binary_emnist_datamodule.py:
```diff
@@ -0,0 +1,180 @@
+        Saves files to data_dir
+        """
+
+        def _prepare_with_splits(split: str):
+            self.dataset_cls(self.data_dir, split=split, train=True, download=True)
+            self.dataset_cls(self.data_dir, split=split, train=False, download=True)
+
+        _prepare_with_splits(self.split)
```
Why not making it simpler?
```suggestion
        self.dataset_cls(self.data_dir, split=split, train=True, download=True)
        self.dataset_cls(self.data_dir, split=split, train=False, download=True)
```

sugatoray replied:
@akihironitta Yes, sure. This completely makes sense. Thank you.

akihironitta commented on pl_bolts/datamodules/emnist_datamodule.py:
```diff
@@ -0,0 +1,240 @@
+
+    Standard EMNIST, train, val, test-splits and transforms
+
+    Transforms::
+
+        emnist_transforms = transform_lib.Compose([
+            transform_lib.ToTensor()
+        ])
```
Could you update the docs here so that it matches with `default_transform`? Similar docs can be found at https://lightning-bolts.readthedocs.io/en/latest/datamodules_vision.html.

sugatoray replied:
@akihironitta Done!

akihironitta commented on pl_bolts/datamodules/emnist_datamodule.py:
```diff
@@ -0,0 +1,240 @@
+            emnist_transforms = transform_lib.Compose([
+                transform_lib.ToTensor(),
+                transform_lib.Normalize(mean=(0.5, ), std=(0.5, )),
+                # TODO: check that EMNIST also uses mean=0.5 and std=0.5
```
Could you resolve this TODO?

sugatoray replied:
@akihironitta I didn't get much time today to work on this one. Let me take a look at it tomorrow/the day after.

sugatoray replied:
@akihironitta Please have a look at [this](https://github.com/PyTorchLightning/lightning-bolts/pull/676#issuecomment-877944327). :star:

akihironitta replied:
@sugatoray It looks great! I just directly updated your branch with a commit adding `emnist_normalization` using the stats (on train sets) you've shared. Mind having a look at the change (955d5b6)? If you find it reasonable, let's use it for `EMNISTDataModule.default_transforms()`.

sugatoray replied:
@akihironitta Thank you for making the update. I haven't checked the docs yet. Did you add these stats in the Binary/EMNISTDataModule docs as well? I think the user should be made aware of these values in the docstring and the example for the default_transforms is where probably making this available could be more helpful for the user. If you have not added this to the docstring yet, would you mind doing that as well?

**A note on choosing `mean = std = 0.5`**: Since the pixel values range between `[0.0, 1.0]`, a transformation such as: `(px - mean) / std` would only scale and shift the pixel values to a range of `[-1.0, 1.0]`. Typically, if you use a model which was already trained with some data and had you used a transformation with some mean and std, later on while using the same model, you would want to use the same mean/std for mimicking the previous data-transformations. However, in this case, the user is going to create a Datamodule and most likely train a model. So, it would not matter if a mean/std of 0.5/0.5 or some other value(s) is/are chosen (although, this may alter the rate of convergence). So, I think that we could be okay with even using 0.5 as the default values for mean and std. But, since we have it calculated already, we might as well provide the mean and std for all the splits in the documentation, in case the user wants to put them to use instead of 0.5. 

**One more thing**: I think this being a statistic of the `EMNIST` dataset, this Mean and SD information should be made part of the `EMNIST_METADATA` and then accessed from there. Whether you use them in the datamodules or not, these are effectively evaluated "facts" about the dataset and so should be part of the `EMNIST_METADATA` and thus will be available in `EMNIST._metadata`.

:bulb: What do you think?

sugatoray replied:
@akihironitta I added the `emnist_normalization()` into emnist_datamodules.py.

akihironitta commented on pl_bolts/datamodules/emnist_datamodule.py:
```diff
@@ -0,0 +1,240 @@
+    ) -> None:
+        """
+        Args:
+            data_dir (string, optional): Where to save/load the data.
```
Could you remove typing in the docs to make it consistent with other docstrings? Types are automatically rendered in the docs :]
```suggestion
            data_dir: Where to save/load the data.
```

sugatoray replied:
@akihironitta sure :D

sugatoray commented:
@akihironitta It seems another error is failing the build. Could you please look into it? As I have not made any changes to `CIFAR10DataModule`, I am not sure what is causing this. :warning:

see: https://github.com/PyTorchLightning/lightning-bolts/pull/676/checks?check_run_id=3031465204

**test_cpcv2**:

`RuntimeError: Given groups=1, weight of size [256, 2048, 1, 1], expected input[2, 1, 32, 32] to have 2048 channels, but got 1 channels instead`

<details>
<summary>
<strong> Expand to see error</strong>
</summary>
<p>

```error
tmpdir = local('/tmp/pytest-of-AzDevOps_azpcontainer/pytest-0/test_cpcv20')
datadir = PosixPath('/__w/4/s/datasets')

    @pytest.mark.skipif(**_MARK_REQUIRE_GPU)
    def test_cpcv2(tmpdir, datadir):
        datamodule = CIFAR10DataModule(data_dir=datadir, num_workers=0, batch_size=2)
        datamodule.train_transforms = CPCTrainTransformsCIFAR10()
        datamodule.val_transforms = CPCEvalTransformsCIFAR10()
    
        model = CPC_v2(
            encoder='mobilenet_v3_small',
            patch_size=8,
            patch_overlap=2,
            online_ft=True,
            num_classes=datamodule.num_classes,
        )
        trainer = Trainer(fast_dev_run=True, default_root_dir=tmpdir)
>       trainer.fit(model, datamodule=datamodule)

tests/models/self_supervised/test_models.py:33: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/home/AzDevOps_azpcontainer/.local/lib/python3.7/site-packages/pytorch_lightning/trainer/trainer.py:460: in fit
    self._run(model)
/home/AzDevOps_azpcontainer/.local/lib/python3.7/site-packages/pytorch_lightning/trainer/trainer.py:758: in _run
    self.dispatch()
/home/AzDevOps_azpcontainer/.local/lib/python3.7/site-packages/pytorch_lightning/trainer/trainer.py:799: in dispatch
    self.accelerator.start_training(self)
/home/AzDevOps_azpcontainer/.local/lib/python3.7/site-packages/pytorch_lightning/accelerators/accelerator.py:96: in start_training
    self.training_type_plugin.start_training(trainer)
/home/AzDevOps_azpcontainer/.local/lib/python3.7/site-packages/pytorch_lightning/plugins/training_type/training_type_plugin.py:144: in start_training
    self._results = trainer.run_stage()
/home/AzDevOps_azpcontainer/.local/lib/python3.7/site-packages/pytorch_lightning/trainer/trainer.py:809: in run_stage
    return self.run_train()
/home/AzDevOps_azpcontainer/.local/lib/python3.7/site-packages/pytorch_lightning/trainer/trainer.py:871: in run_train
    self.train_loop.run_training_epoch()
/home/AzDevOps_azpcontainer/.local/lib/python3.7/site-packages/pytorch_lightning/trainer/training_loop.py:499: in run_training_epoch
    batch_output = self.run_training_batch(batch, batch_idx, dataloader_idx)
/home/AzDevOps_azpcontainer/.local/lib/python3.7/site-packages/pytorch_lightning/trainer/training_loop.py:738: in run_training_batch
    self.optimizer_step(optimizer, opt_idx, batch_idx, train_step_and_backward_closure)
/home/AzDevOps_azpcontainer/.local/lib/python3.7/site-packages/pytorch_lightning/trainer/training_loop.py:442: in optimizer_step
    using_lbfgs=is_lbfgs,
/home/AzDevOps_azpcontainer/.local/lib/python3.7/site-packages/pytorch_lightning/core/lightning.py:1403: in optimizer_step
    optimizer.step(closure=optimizer_closure)
/home/AzDevOps_azpcontainer/.local/lib/python3.7/site-packages/pytorch_lightning/core/optimizer.py:214: in step
    self.__optimizer_step(*args, closure=closure, profiler_name=profiler_name, **kwargs)
/home/AzDevOps_azpcontainer/.local/lib/python3.7/site-packages/pytorch_lightning/core/optimizer.py:134: in __optimizer_step
    trainer.accelerator.optimizer_step(optimizer, self._optimizer_idx, lambda_closure=closure, **kwargs)
/home/AzDevOps_azpcontainer/.local/lib/python3.7/site-packages/pytorch_lightning/accelerators/accelerator.py:329: in optimizer_step
    self.run_optimizer_step(optimizer, opt_idx, lambda_closure, **kwargs)
/home/AzDevOps_azpcontainer/.local/lib/python3.7/site-packages/pytorch_lightning/accelerators/accelerator.py:336: in run_optimizer_step
    self.training_type_plugin.optimizer_step(optimizer, lambda_closure=lambda_closure, **kwargs)
/home/AzDevOps_azpcontainer/.local/lib/python3.7/site-packages/pytorch_lightning/plugins/training_type/training_type_plugin.py:193: in optimizer_step
    optimizer.step(closure=lambda_closure, **kwargs)
/usr/local/lib/python3.7/dist-packages/torch/autograd/grad_mode.py:15: in decorate_context
    return func(*args, **kwargs)
/usr/local/lib/python3.7/dist-packages/torch/optim/adam.py:62: in 
```

</p>
</details>

sugatoray commented:
See here: [emnist_summary.txt](https://github.com/PyTorchLightning/lightning-bolts/files/6798428/emnist_summary.txt) for the mean and SD calculated for each split in EMNIST.

```txt
{ 'balanced': { 'split': 'balanced',
                'summary': { 'test': { 'mean': 0.17552242362752873,
                                       'num_units': 18800,
                                       'std': 0.33343913740623093,
                                       'unit_shape': (1, 28, 28)},
                             'train': { 'mean': 0.17510417221708502,
                                        'num_units': 112800,
                                        'std': 0.3332070017067981,
                                        'unit_shape': (1, 28, 28)}}},
  'byclass': { 'split': 'byclass',
               'summary': { 'test': { 'mean': 0.1736120948240831,
                                      'num_units': 116323,
                                      'std': 0.33162168261272984,
                                      'unit_shape': (1, 28, 28)},
                            'train': { 'mean': 0.17359222670426913,
                                       'num_units': 697932,
                                       'std': 0.33162134741938604,
                                       'unit_shape': (1, 28, 28)}}},
  'bymerge': { 'split': 'bymerge',
               'summary': { 'test': { 'mean': 0.1735874978201131,
                                      'num_units': 116323,
                                      'std': 0.33162838805442035,
                                      'unit_shape': (1, 28, 28)},
                            'train': { 'mean': 0.17359632632958918,
                                       'num_units': 697932,
                                       'std': 0.33161854660044826,
                                       'unit_shape': (1, 28, 28)}}},
  'digits': { 'split': 'digits',
              'summary': { 'test': { 'mean': 0.17353817701339722,
                                     'num_units': 40000,
                                     'std': 0.33195470327000776,
                                     'unit_shape': (1, 28, 28)},
                           'train': { 'mean': 0.17325182375113168,
                                      'num_units': 240000,
                                      'std': 0.33163191505859535,
                                      'unit_shape': (1, 28, 28)}}},
  'letters': { 'split': 'letters',
               'summary': { 'test': { 'mean': 0.1721735894221526,
                                      'num_units': 20800,
                                      'std': 0.32958498872403497,
                                      'unit_shape': (1, 28, 28)},
                            'train': { 'mean': 0.17222730561708793,
                                       'num_units': 124800,
                                       'std': 0.33091591285642147,
                                       'unit_shape': (1, 28, 28)}}},
  'mnist': { 'split': 'mnist',
             'summary': { 'test': { 'mean': 0.17345932738780975,
                                    'num_units': 10000,
                                    'std': 0.3316542240944672,
                                    'unit_shape': (1, 28, 28)},
                          'train': { 'mean': 0.17330445484320323,
                                     'num_units': 60000,
                                     'std': 0.33169403605816716,
                                     'unit_shape': (1, 28, 28)}}}}
```

akihironitta commented:
> @akihironitta It seems another error is failing the build. Could you please look into it? As I have not made any changes to `CIFAR10DataModule`, I am not sure what is causing this. :warning:

You're right it seems like the error is unrelated to the change in this PR, so we can just ignore it here for now and fix it in another PR later.

sugatoray commented:
@akihironitta All suggested changes were taken into account. The only one that is still open is about [`_emnist_split_exists`](https://github.com/PyTorchLightning/lightning-bolts/pull/676#discussion_r665078561). Please have a look at the code and let me know if you have any other suggestions.

akihironitta reviewed (approved):
LGTM!

**Description of the change**: Introduced a new arg `strict_val_split: bool` to use the balanced validation set defined in the EMNIST paper.

(always open for any discussions/comments on the API (`strict_val_split`) :])

sugatoray reviewed (commented):
@akihironitta Looks good to me. Thank you, for working on it. I am not sure, if you were waiting for any feedback from me. It seems that this has not been reviewed by other reviewers since your approval. Please let me know if there is something I need to do.

akihironitta commented:
@sugatoray Hi, thanks for having a look again!

@PyTorchLightning/core-bolts Would you mind having a look at the changes?

tchaton commented on pl_bolts/datamodules/emnist_datamodule.py:
```diff
@@ -0,0 +1,221 @@
+        data_dir: Optional[str] = None,
+        split: str = 'mnist',
+        val_split: Union[int, float] = 0.2,
+        num_workers: int = 16,
```
`num_workers` seem quite high for a default.

sugatoray replied:
@tchaton These are the same defaults as that in `pl_bolts/datamodules/mnist_datamodule.py`. Although, I personally feel that `num_workers = 16` as a default is quite high. I would suggest to make it `2` instead of `16` (since, Google Colab, Kaggle, most laptops provide at least two cores).

@akihironitta @Borda What do you think about making this default `num_workers = 2` (instead of `16`)?

```python
class MNISTDataModule(VisionDataModule):
    """ doctring here...
    """
    name = "mnist"
    dataset_cls = MNIST
    dims = (1, 28, 28)

    def __init__(
        self,
        data_dir: Optional[str] = None,
        val_split: Union[int, float] = 0.2,
        num_workers: int = 16,
        normalize: bool = False,
        batch_size: int = 32,
        seed: int = 42,
        shuffle: bool = False,
        pin_memory: bool = False,
        drop_last: bool = False,
        *args: Any,
        **kwargs: Any,
    ) -> None:
```

akihironitta replied:
Opened a quick PR #701 for this suggestion, so resolving this thread.

tchaton commented on pl_bolts/datamodules/emnist_datamodule.py:
```diff
@@ -0,0 +1,221 @@
+        normalize: bool = False,
+        batch_size: int = 32,
+        seed: int = 42,
+        shuffle: bool = False,
+        pin_memory: bool = False,
```
```suggestion
        shuffle: bool = True,
        pin_memory: bool = True,
```

sugatoray replied:
@tchaton These are the same defaults used in other existing datamodules.

```python
num_workers: int = 16,
shuffle: bool = False,
pin_memory: bool = False,
```

Some of the existing datamodules where the above defaults are used currently.

- [x] `binary_mnist_datamodule`
- [x] `cifar10_datamodule`
- [x] `mnist_datamodule`
- [x] `cityscapes_datamodule`
- [x] `fashion_mnist_datamodule`
- [x] `imagenet_datamodule`

So, in the light of this, I would suggest that the same defaults, for at least `shuffle` and `pin_memory`, be used for `emnist_datamodule` and `binary_emnist_datamodule`. If any changes must be made, let that be part of a separate issue and make the changes to all of them in a separate PR.

cc: @akihironitta @Borda

akihironitta replied:
Opened a quick PR #701 for this suggestion, so resolving this thread.

tchaton commented on pl_bolts/datamodules/emnist_datamodule.py:
```diff
@@ -0,0 +1,221 @@
+                self.data_dir, split=self.split, train=False, transform=test_transforms, **self.EXTRA_ARGS
+            )
+
+    def default_transforms(self) -> Callable:
```
This can be simplified

akihironitta replied:
Should be addressed now. Would you mind having a look again?

akihironitta commented:
The failing tests are irrelevant to the change in this PR.

<details><summary>pytest summary at c6adbae</summary>

```
=========================== short test summary info ============================
FAILED tests/callbacks/verification/test_base.py::test_verification_base_get_input_array[device0]
FAILED tests/callbacks/verification/test_batch_gradient.py::test_batch_gradient_verification_pl_module[device0-True]
FAILED tests/callbacks/verification/test_batch_gradient.py::test_batch_gradient_verification_pl_module[device0-False]
FAILED tests/callbacks/verification/test_batch_gradient.py::test_batch_gradient_verification_callback[0]
FAILED tests/models/rl/test_scripts.py::test_cli_run_rl_dqn[ --env PongNoFrameskip-v4 --max_steps 10 --fast_dev_run 1 --warm_start_size 10 --n_steps 2 --batch_size 10]
FAILED tests/models/rl/test_scripts.py::test_cli_run_rl_reinforce[ --env CartPole-v0 --max_steps 10 --fast_dev_run 1 --batch_size 10]
FAILED tests/models/rl/test_scripts.py::test_cli_run_rl_vanilla_policy_gradient[ --env CartPole-v0 --max_steps 10 --fast_dev_run 1 --batch_size 10]
= 7 failed, 288 passed, 23 skipped, 1 xfailed, 532 warnings in 1053.39s (0:17:33) =
```

</details>

akihironitta commented:
@PyTorchLightning/core-bolts Could you have another look?
